File: comparison/bobble/resources/melchior/application.hs
module Comparisons.Bobble where

import Control.Category hiding ((>>>), (<<<))
import Prelude hiding ((.), id)

import Melchior.Control
import Melchior.Dom
import Melchior.Dom.Drawing
import Melchior.Dom.Selectors
import Melchior.EventSources.Keyboard
import Melchior.Time

main :: IO ()
main = runDom setupBobble

frame = 17 -- ~16.66ms / 60fps

data Game   = Game Player Player Ball
data Player = Player { x :: Int, y :: Int, vx :: Float, vy :: Float, score :: Int }
data Ball   = Ball { xb :: Int, yb :: Int, vxb :: Float, vyb :: Float }

setupBobble html = do
  canvas <- Dom $ assuredly $ select (canvases . byId "canvas" . from) html
  let context = contextOf canvas
      game = Game (Player 50 350 0 0 0) (Player 650 350 0 0 0) (Ball 350 50 4 2)
      keys = keyCode $ keyDownSignal (toElement document)
  play game context keys

play :: Game -> Context -> Signal Int -> Dom ()
play g context keys = terminate (after frame)
                      (\_ -> do
                          drawBackground context
                          h <- return $ scoreG $ collide $ move g $ takes keys
                          drawGameElements h context
                          let Dom io = play h context keys in io)

scoreG :: Game -> Game
scoreG g@(Game _ _ b) = if yb b > 360 then reset $ scoreP (xb b < 375) g else g

reset :: Game -> Game
reset (Game p1 p2 _) = Game p1 p2 (Ball 350 50 4 4)

scoreP :: Bool -> Game -> Game
scoreP False (Game (Player x y vx vy s) p b) = Game (Player x y vx vy (s+1)) p b
scoreP True  (Game p (Player x y vx vy s) b) = Game p (Player x y vx vy (s+1)) b

drawBackground :: Context -> IO ()
drawBackground c = do
  let Dom io = rectangle 0 350 750 400 (fillStyle c "#999") in io
  let Dom io = rectangle 0 0 750 350 (fillStyle c "#FFF") in io
  let Dom io = rectangle 375 320 1 60 (fillStyle c "#3F3") in io

drawGameElements :: Game -> Context -> IO ()
drawGameElements (Game p1 p2 b) c = do
  drawScores (score p1) (score p2) c
  drawPlayer p1 (fillStyle c red)
  drawPlayer p2 (fillStyle c blue)
  drawBall b (fillStyle c green)

drawScores :: Int -> Int -> Context -> IO ()
drawScores p1 p2 c = do
  let Dom io = text (show p1) 25 50 (fontStyle c "40pt Helvetica") in io
  let Dom io = text (show p2) 650 50 (fontStyle c "40pt Helvetica") in io

drawPlayer :: Player -> Context -> IO ()
drawPlayer p c = let Dom io = circle (x p) (y p) 25 c in io

drawBall :: Ball -> Context -> IO ()
drawBall b c = let Dom io = circle (xb b) (yb b) 10 c in io

collide :: Game -> Game
collide (Game p1 p2 b) = Game (collideP p1) (collideP p2) (collideBWall $ collideB p1 p2 b)

collideP :: Player -> Player
collideP p = if ((abs $ (x p) - 375) < 25) || ((x p) < 25) || ((x p) > 700)
             then Player (x p) (y p) ((vx p)* (-1)) (vy p) (score p)
             else p

collideB :: Player -> Player -> Ball -> Ball
collideB p1 p2 b = collideWith p1 $ collideWith p2 b

collideBWall :: Ball -> Ball
collideBWall b = if (xb b) < 10 || (xb b) > 740 then Ball (xb b) (yb b) ((vxb b)*(-1)) (vyb b) else b

collideWith :: Player -> Ball -> Ball
collideWith p b = if ((x p - (xb b))^2 + (y p - (yb b))^2) < (25^2) then collision p b else b

collision :: Player -> Ball -> Ball
collision p b = Ball ((xb b) + 2*(round $ fst v')) ((yb b) + 2*(round $ snd v')) (fst v') (snd v')
  where
    n  = normalisedVectorBetween (x p, y p) ((xb b), (yb b))
    a1 = dot ((vxb b), (vyb b)) n
    a2 = dot (vx p, vy p) n
    op = (2.0 * (a2 - a1))/4
    v' = prod (-0.95) $ diff ((vxb b), (vyb b)) (prod op $ prod 2 n)

normalisedVectorBetween :: (Int, Int) -> (Int, Int) -> (Float, Float)
normalisedVectorBetween a b = (dx/d, dy/d)
                    where
                      d  = sqrt $ (dx * dx) + (dy * dy)
                      dx = fromIntegral (fst a - fst b)
                      dy = fromIntegral (snd a - snd b)

dot :: (Float, Float) -> (Float, Float) -> Float
dot a b = (fst a * fst b) + (snd a * snd b)

prod :: Float -> (Float, Float) -> (Float, Float)
prod s v = ((s*fst v), (s*snd v))

diff :: (Float, Float) -> (Float, Float) -> (Float, Float)
diff a b = ((fst a - fst b), (snd a - snd b))

move :: Game -> Int -> Game
move (Game p1 p2 b) i = if i > 40
                        then Game (gravitate $ moveP p1 i) (gravitate $ moveP p2 (-1)) (moveB b)
                        else Game (gravitate $ moveP p1 (-1)) (gravitate $ moveP p2 i) (moveB b)
dv = 5.0
moveP :: Player -> Int -> Player
moveP p i = if (i == 37 || i == 65)
            then Player ((x p) + (round (vx p))) (y p) ((vx p) - dv) (vy p) (score p)
            else if (i == 38 || i == 87)
                 then (if (y p) == 350
                       then Player (x p) ((y p) + (round (vy p))) (vx p) ((vy p) - 5*dv) (score p)
                       else Player ((x p) + (round (vx p))) ((y p) + (round (vy p))) ((vx p)*0.9) ((vy p)*0.9) (score p))
                 else
                   if (i == 39 || i == 68)
                   then Player ((x p) + (round (vx p))) (y p) ((vx p) + dv) (vy p) (score p)
                   else Player ((x p) + (round (vx p))) ((y p) + (round (vy p))) ((vx p)*0.9) ((vy p)*0.9) (score p)

gravitate (Player x y vx vy s) = Player x (if y > 350 then 350 else y) vx (vy+0.5) s

moveB :: Ball -> Ball
moveB b = Ball ((xb b)+(round (vxb b))) ((yb b)+(round (vyb b))) (vxb b) ((vyb b)+0.25)

File: comparison/bobble/resources/js/application.js
!function () {

    function main () {
        var canvas = document.getElementById("canvas")
          , context = canvas.getContext("2d")

        playGame(context, new Player(50, 350), new Player(650, 350), new Ball())
    }

    function Player (x, y, vx, vy, score) {
        this.x = x || 50,
        this.y = y || 350,
        this.vx = vx || 0,
        this.vy = vy || 0,
        this.score = score || 0
    }

    function Ball (x, y, vx, vy) {
        this.x = x || 350,
        this.y = y || 50,
        this.vx = vx || 4,
        this.vy = vy || 2
    }

    function playGame(ctx, p1, p2, ball) {
        registerKeyListeners(p1, p2)
        startGameLoop(ctx, p1, p2, ball)
    }

    function registerKeyListeners(p1, p2) {
        var dv = 5.0
        document.addEventListener("keydown", function (e) {
            function movePlayer(p, l, u, r) {
                if (e.keyCode == l) {
                    p.x += p.vx
                    p.vx -= dv
                } else if (e.keyCode == u) {
                    if (p.y >= 350) {
                        p.vy -= 5 * dv
                        p.y += p.vy
                    } else {
                        p.x += p.vx
                        p.y += p.vy
                    }
                } else if (e.keyCode == r) {
                    p.x += p.vx
                    p.vx += dv
                }
            }
            movePlayer(p1, 65, 87, 68)
            movePlayer(p2, 37, 38, 39)
        })
    }

    function startGameLoop (ctx, p1, p2, ball) {
        !function loop(ctx, p1, p2, ball) {
            drawBackground(ctx)
            move(p1, p2, ball)
            collide(p1, p2, ball)
            score(p1, p2, ball)
            drawGame(ctx, p1, p2, ball)
            setTimeout(function () {
                loop(ctx, p1, p2, ball)
            }, 17)
        }(ctx, p1, p2, ball)
    }

    function move (p1, p2, ball) {
        function lowerBound(p) {
            if (p.y > 350) p.y = 350
            p.vy += 0.5
        }
        function tick(p) {
            p.x += p.vx
            p.y += p.vy
            p.vx *= 0.9
            p.vy *= 0.9
        }
        lowerBound(p1)
        lowerBound(p2)
        tick(p1)
        tick(p2)
        ball.x += ball.vx
        ball.y += ball.vy
        ball.vy += 0.25
    }

    function collide(p1, p2, ball) {
        collideWithWalls(p1, 25)
        collideWithWalls(p2, 25)
        collideWithWalls(ball, 10)
        collideWithBall(p1, ball)
        collideWithBall(p2, ball)
    }

    function collideWithWalls (x, r) {
        if(x.x < r || x.x > (700 - r) || r > 10 && Math.abs(x.x-375) < r) {
            x.vx *= -1
            x.x += x.vx
        }
    }

    function collideWithBall (p, b) {
        if( ((p.x - b.x) * (p.x - b.x)) + ((p.y - b.y) * (p.y - b.y)) < (25 * 25)) {
            var n = normalisedVectorBetween(p, b)
            var a1 = dot({x:b.vx, y:b.vy}, n)
            var a2 = dot({x:p.vx, y:p.vy}, n)
            var op = (2*(a2-a1))/4
            var v = prod(-0.99, diff({x:b.vx, y:b.vy}, prod(op, prod(2, n))))
            b.x += 2*v.x
            b.y += 2*v.y
            b.vx += v.x
            b.vy += v.y
        }
    }

    function prod(scalar, vector) {
        return {x: scalar*vector.x, y: scalar*vector.y }
    }

    function diff(a, b) {
        return {x: a.x - b.x, y: a.y - b.y}
    }

    function dot (p, n) {
        return p.x * n.x + p.y * n.y
    }

    function normalisedVectorBetween(p, b) {
        var dx = p.x - b.x, dy = p.y - b.y
        var d = Math.sqrt(dx*dx + dy*dy)
        return { x:dx/d, y: dy/d }
    }

    function score(p1, p2, ball) {
        if(ball.y > 360) {
            if(ball.x < 375) p2.score += 1
            else p1.score += 1
            ball.x = 350
            ball.y = 50
            ball.vx = 4
            ball.vy = 2
        }
    }

    function drawGame(ctx, p1, p2, ball) {
        ctx.font = "40pt Helvetica"
        ctx.fillText(p1.score, 25, 50)
        ctx.fillText(p2.score, 650, 50)
        ctx.fillStyle = "#FF0000"
        ctx.beginPath()
        ctx.arc(p1.x, p1.y, 25, 0, 2*Math.PI, false)
        ctx.fill()
        ctx.stroke()
        ctx.fillStyle = "#0000FF"
        ctx.beginPath()
        ctx.arc(p2.x, p2.y, 25, 0, 2*Math.PI, false)
        ctx.fill()
        ctx.stroke()
        ctx.fillStyle = "#00FF00"
        ctx.beginPath()
        ctx.arc(ball.x, ball.y, 10, 0, 2*Math.PI, false)
        ctx.fill()
        ctx.stroke()
    }

    function drawBackground(ctx) {
        ctx.fillStyle = "#999"
        ctx.fillRect(0, 350, 750, 400)
        ctx.fillStyle = "#FFF"
        ctx.fillRect(0, 0, 750, 350)
        ctx.fillStyle = "#3F3"
        ctx.fillRect(375, 320, 1, 60)
    }

    main()
} ()